src/librustc_mir_build/lints.rs

```diff
@@ -114,6 +114,10 @@ impl<'mir, 'tcx> TriColorVisitor<&'mir Body<'tcx>> for Search<'mir, 'tcx> {
             | TerminatorKind::Unreachable
             | TerminatorKind::Yield { .. } => ControlFlow::Break(NonRecursive),
 
+            // FIXME(Amanieu): I am not 100% sure about this, but it triggers
+            // a spurious warning otherwise.
+            TerminatorKind::InlineAsm { .. } => ControlFlow::Break(NonRecursive),
```

Comment:
Could someone more familiar with the infinite recursion lint have a look at this? Basically the issue is that functions returning `!` still have an unwind edge that eventually reaches one of the terminating cases above (nolandingpads only happens later on in MIR). However a `noreturn` asm does not have this unwind edge which means that it would spuriously trigger this lint.

Reply:
cc @ecstatic-morse who recently rewrote that lint in #70973.

Reply:
That comment should proabably read "return control flow to the caller or abort". Seems like `noreturn` ASM should be treated like `Abort` and regular ASM should be treated like `Call`?

Reply:
I was confused because the way the lint currently works: all calls are treated as possibly divergent due to the unwind edge. For example this code does not trigger the lint:
```rust
fn bar() {}

fn foo() {
    bar();
    foo();
}
```

Reply:
That's true, although that's just a shortcoming of the analysis since we (I think) don't know that `bar` cannot panic at this point in the MIR pipeline, so our CFG traversal eventually sees a `TerminatorKind::Resume`. If we actually do know this, I'd be happy to fix it.

Does this help or are things still unclear? Can all inline assembly possibly abort? Or just `noreturn`? I would assume it's the former.

Reply:
I suppose your question is more like: If we assume a call to `bar` could possibly panic, so the program *technically* does not recurse infinitely, shouldn't we still emit the lint in this case?

Reply:
Yes, that is what I am saying. We could ignore all call unwind edges for the purposes of the lint and only return `ControlFlow::Break(NonRecursive)` for diverging functions.

But anyways, let's leave that for a separate issue/PR. I've updated the code in this PR to only break for `noreturn` asm.